Fix pkg.install when pkg already installed

twangboy opened the pull request:
### What does this PR do?
Fixes an issue with pkg.install on Windows where the `current: <version>` was being returned if you passed the same version that was installed. Correct behavior, when there are no changes, is to return an empty dictionary.

Adds tests to test valid output.

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/issues/48931

### Tests written?
Yes

### Commits signed with GPG?
Yes

twangboy committed:
Fix issues where current is not returned

damon-atkins reviewed (changes requested):
Only place return results should be played with is a schedule job which is not waited for.

rallytime commented:
@twangboy Can we get a test written for this so it does not regress again?

damon-atkins commented:
Hi Guys this is not the correct fix. I have updated the issue #48931

twangboy commented on salt/modules/win_pkg.py:
```diff
@@ -1217,31 +1217,36 @@ def install(name=None, refresh=False, pkgs=None, **kwargs):
-                          '%s', version_num, pkg_name)
+                log.debug('"%s" version "%s" is already installed',
+                          pkg_name, old[pkg_name][0])
+                ret[pkg_name] = {'current': old[pkg_name][0]}
```
This is required. It is the fix to the issue. If you don't pass a version number and the package is already installed it needs to return `current: <version number>`. Without this line it returns nothing.

twangboy replied:
I don't know what you mean by the salt standard.

damon-atkins replied:
It should return no results for no changes. This is the standard across all the other platforms.
Please look at how the other implementation work.

The expected salt behaviour is the following example. 
```
# salt-call -l info  pkg.version curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    7.59.0-5.fc28
```
The above is the current version. So below will show no changes.  i.e. no need output the current version.
```
# salt-call -l info  pkg.install curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'systemd-run', u'--scope', u'dnf', u'-y', u'--best', u'--allowerasing', u'install', u'curl'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    ----------
```


The following is a change example:
```
# salt-call -l info  pkg.version curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    7.59.0-2.fc28

# salt-call -l info  pkg.install curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'systemd-run', u'--scope', u'dnf', u'-y', u'--best', u'--allowerasing', u'install', u'curl'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    ----------
    curl:
        ----------
        new:
            7.59.0-5.fc28
        old:
            7.59.0-2.fc28
    libcurl:
        ----------
        new:
            7.59.0-5.fc28
        old:
            7.59.0-2.fc28
```

The current version of win_pkg matches this behaviour.  Which all the platforms feed into states/pkg.py

win_pkg needs to move closer to the behaviour of the other platforms. For example it needs to raise more CommandExecutionError  where errors come out as comments under states/pkg.py

For example in states/pkg.py under def installed is:
```
        except CommandExecutionError as exc:
            ret = {'name': name, 'result': False}
            if exc.info:
                # Get information for state return from the exception.
                ret['changes'] = exc.info.get('changes', {})
                ret['comment'] = exc.strerror_without_changes
            else:
                ret['changes'] = {}
                ret['comment'] = ('An error was encountered while installing '
                                  'package(s): {0}'.format(exc))
            if warnings:
                ret.setdefault('warnings', []).extend(warnings)
            return ret
```

damon-atkins replied:
Hi, It should return no results for no changes, as this is the standard across all the other platforms. Please look at how the other implementation work.

The following example shows the salt expected behaviour: 
```
# salt-call -l info  pkg.version curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    7.59.0-5.fc28
```
The above is the current version. So below will show no changes.  i.e. no need output the current version.
```
# salt-call -l info  pkg.install curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'systemd-run', u'--scope', u'dnf', u'-y', u'--best', u'--allowerasing', u'install', u'curl'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    ----------
```


The following is a change example:
```
# salt-call -l info  pkg.version curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    7.59.0-2.fc28

# salt-call -l info  pkg.install curl
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'systemd-run', u'--scope', u'dnf', u'-y', u'--best', u'--allowerasing', u'install', u'curl'] in directory '/root'
[INFO    ] cmdmod._run Executing command [u'rpm', u'-qa', u'--queryformat', u'%{NAME}_|-%{EPOCH}_|-%{VERSION}_|-%{RELEASE}_|-%{ARCH}_|-(none)_|-%{INSTALLTIME}'] in directory '/root'
local:
    ----------
    curl:
        ----------
        new:
            7.59.0-5.fc28
        old:
            7.59.0-2.fc28
    libcurl:
        ----------
        new:
            7.59.0-5.fc28
        old:
            7.59.0-2.fc28
```

The current version of win_pkg matches this behaviour (mostly more work is needed).  The results then feed into "changes" under states/pkg.py

win_pkg needs to move closer to the behaviour of the other platforms. For example it needs to raise more CommandExecutionError  where errors come out as comments under states/pkg.py

For example in states/pkg.py under def installed is:
```
        except CommandExecutionError as exc:
            ret = {'name': name, 'result': False}
            if exc.info:
                # Get information for state return from the exception.
                ret['changes'] = exc.info.get('changes', {})
                ret['comment'] = exc.strerror_without_changes
            else:
                ret['changes'] = {}
                ret['comment'] = ('An error was encountered while installing '
                                  'package(s): {0}'.format(exc))
            if warnings:
                ret.setdefault('warnings', []).extend(warnings)
            return ret
```

damon-atkins replied:
Windows
```
# salt win10test state.high {'grepwin':{'pkg':['installed']}}
win10test:
    ----------
    pkg_|-grepwin_|-grepwin_|-installed:
        ----------
        __id__:
            grepwin
        __run_num__:
            0
        __sls__:
            None
        changes:
            ----------
            grepwin:
                ----------
                new:
                    1.6.682
                old:
        comment:
            The following packages were installed/updated: grepwin
        duration:
            28764.621
        name:
            grepwin
        result:
            True
        start_time:
            02:49:50.644832

# salt win10test state.high {'grepwin':{'pkg':['installed']}}
win10test:
    ----------
    pkg_|-grepwin_|-grepwin_|-installed:
        ----------
        __id__:
            grepwin
        __run_num__:
            0
        __sls__:
            None
        changes:
            ----------
        comment:
            All specified packages are already installed
        duration:
            28122.063
        name:
            grepwin
        result:
            True
        start_time:
            02:51:30.499313
```
Unix/Linux
```
# salt master state.high {'curl':{'pkg':['installed']}}
master:
    ----------
    pkg_|-curl_|-curl_|-installed:
        ----------
        __id__:
            curl
        __run_num__:
            0
        __sls__:
            None
        changes:
            ----------
        comment:
            All specified packages are already installed
        duration:
            5437.673
        name:
            curl
        result:
            True
        start_time:
            02:53:54.484794
```

twangboy commented on salt/modules/win_pkg.py:
```diff
@@ -1217,31 +1217,36 @@ def install(name=None, refresh=False, pkgs=None, **kwargs):
+                ret[pkg_name] = {'current': old[pkg_name][0]}
                 continue
-            # following can be version number or latest or Not Found
+            # Get the most recent version number available from winrepo.p
```
I'll clarify this

twangboy commented on salt/modules/win_pkg.py:
```diff
@@ -1217,31 +1217,36 @@ def install(name=None, refresh=False, pkgs=None, **kwargs):
             continue
         # If version number not installed, is the version available?
-        elif version_num != 'latest' and version_num not in pkginfo:
+        elif version_num.lower() != 'latest' and version_num not in pkginfo:
```
`version_num` is populated with the contents of the `version` parameter passed either in a state file or on the CLI. Since it is user input, I added the `.lower()` just to make sure.
Later in the code, if `version` is not specified, `version_num` is then populated with the results of the `_get_latest_pkg_version(pkginfo)` command.
There aren't that many packages that use the `latest` option in the Software Definition file. But, it probably wouldn't hurt to run `.lower()` on all of them to account for inconsistencies.

damon-atkins replied:
Note: states/pkg.py only supports `latest`  There is no `.lower()` used in states/pkg.py.
Allowing it on the cli or in the sls package definitions might confusion the topic as people may try to use it with `pkg.installed` state. Hence I would suggest a reminder in the doco.  And maybe an error in a package definition file reported by refresh_db. Educate the right way.

twangboy replied:
So, you're suggesting to let the function fail if the user passes `Latest` instead of `latest`?

twangboy replied:
If `Latest` is passed in the state system, wouldn't it be caught here?

twangboy commented:
This is the fix to THIS issue

twangboy commented:
@rallytime I'll write some tests

twangboy committed:
Clarify code comment

twangboy committed:
Add tests for pkg.install output

twangboy committed:
Make the tests run on Linux

damon-atkins commented:
Hi @terminalmage 

If pkg.install or pkg.remove end up not making any changes to the target system they return an empty dict indicating no changes. This is the salt standard for pkg.install and pkg.remove is to return differences in the packages and their version.  Current release of 2016.11, 2017 and 2018 behaviour this way.

The PR will move win_pkg away from that standard, cause states/pkg.py to get data it would not normal expect.    i.e. `current: version number` to show what version is currently installed.

This PR will change the output to
```
salt-call pkg.version
local:
    2.2.6
salt-call pkg.install vlc version=2.2.6
local:
    ----------
    vlc:
        ----------
        current:
            2.2.6
```
Instead of standard empty dict for no changes in software or version
```
salt-call pkg.install vlc version=2.2.6
local:
    ----------
```

Which will result in output like
```
win10test:
----------
          ID: vlc
    Function: pkg.installed
      Result: True
     Comment: Package vlc=2.2.6 was reinstalled.
     Started: 12:58:06.916154
    Duration: 19869.135 ms
     Changes:
              ----------
              vlc:
                  ----------
                  current:
                      2.2.6
```
vs
```
    ID: curl
    Function: pkg.installed
      Result: True
     Comment: Package curl=7.59.0-5.fc28 was reinstalled.
     Started: 13:06:09.163772
    Duration: 71738.376 ms
     Changes:
              ----------
              curl:
                  ----------
                  new:
                      7.59.0-5.fc28
                  old:
                      7.59.0-5.fc28
```

damon-atkins commented on salt/modules/win_pkg.py:
```diff
@@ -1217,31 +1217,37 @@ def install(name=None, refresh=False, pkgs=None, **kwargs):
             continue
 
-        version_num = options.get('version', '')
+        version_num = options.get('version')
```
Are you saying that version will always exist under options and hence there is no need for a default value to prevent an error.

twangboy replied:
No, just using the default `None`

twangboy committed:
return empty dict on no changes

twangboy committed:
Fix typo in code comment

twangboy commented:
@damon-atkins I have modified the PR to return an empty dict when there are no changes. I have also added tests to ensure correct returns in those scenarios. Thank you for educating me on proper salt behavior in regards to the pkg modules. Elsewhere in salt, execution module outputs are not standardized, by design apparently (I asked around). But in the case of the pkg system, there is some standardization. Thanks for keeping me straight.

twangboy committed:
Remove current: version for latest as well

damon-atkins reviewed (changes requested):
Some more suggestions

twangboy committed:
remove .lower(), fix debug messages

damon-atkins commented on salt/modules/win_pkg.py:
```diff
@@ -1217,28 +1211,33 @@ def install(name=None, refresh=False, pkgs=None, **kwargs):
                 continue
-            # following can be version number or latest or Not Found
+            # Get the most recent version number available from winrepo.p
+            # May also return `latest` or an empty string
```
or 'Not Found'    a package definition can contain Not Found to match when the version string in the app can not be found.

damon-atkins reviewed (approved):
Subject to comment update. Thanks
